### PR TITLE
Define resource semantic convention stability and justification

### PR DESCRIPTION
**Pull request by jsuereth**
- Define what must remain stable in resource detection
- Outline issues with changes to resource attributes and the associated identity of telemetry streams

NOTE: This is incomplete and posted for discussion purposes only.

### CODE REVIEW

File: 0162-semantic-convention-resource-stability.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  all or nothing fashion. if a resource detector produced attributes XYZ for
+  a given compute, it should continue to do so for compatible versions.
+- Resource attributes (e.g. `cloud.provider`) may have new possible values
+  defined as long as there is zero overlap with how the previous values were
+  computed.  For example, `acme-cloud` would never be inferred where previously
+  `aws` was inferred.
+  *Note: if `aws` was inferred incorrectly on "Acme cloud" previously, that would not be a violation.*
+- Adding new Resource attribute semantic conventions is allowed, however
+  existing resource detectors MUST NOT use them, only new detectors would be
+  allowed.
```
Comment (tedsuo) [inline 0162-semantic-convention-resource-stability.md:37]: This last point is a bit vague. What counts as a new detector? If an attribute is added to the K8s resource definition, how would it be handled?

File: 0162-semantic-convention-resource-stability.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+A resource detector providing K8s node information MUST produce the following.
+
+**type:** `k8s.node`
+
+**Description:** A Kubernetes Node.
+
+<!-- semconv k8s.node -->
+| Attribute  | Type | Description  | Examples  | Required |
+|---|---|---|---|---|
+| `k8s.node.name` | string | The name of the Node. | `node-1` | *No* |
```
Comment (tedsuo) [inline 0162-semantic-convention-resource-stability.md:73]: How does the stability requirement interact with optional attributes?

File: 0162-semantic-convention-resource-stability.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+- Resource semantic conventions will need to be defined at the
+  `Resource Detector` level to ensure stability of identity of resources.
+- Attribute names and existing values cannot be renamed within stable versions
+  (pending OTEP #161)
+
+### New Semantic Conventions
+
+Going forward, Resource semantic conventions will be outlined at a "Resource
+detection" level.  While users can merge together resource detectors, each
+detector will provide a stable set of Resource attributes and labels.
```
Comment (tedsuo) [inline 0162-semantic-convention-resource-stability.md:57]: How do resource detectors differ from the current namespacing regime for resource attributes? Is `k8s.*` a resource detector? Is `k8s.cluster.*` a resource detector?

File: 0162-semantic-convention-resource-stability.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# Resource Semantic Convention Stabiltiy Requirements
+
+Outline what changes are "safe" to perform for resource semantic conventions
+and which require major version bumps.
```
Comment (tigrannajaryan) [inline 0162-semantic-convention-resource-stability.md:4]: I am a bit confused as to why we need this document.

I believe that now that starting from 1.4.0 we have Telemetry Schemas, the Schemas defines what changes are safe. If the change is representable as version change in the Schema then it is safe, everything else is unsafe.

File: 0162-semantic-convention-resource-stability.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  computed.  For example, `acme-cloud` would never be inferred where previously
+  `aws` was inferred.
+  *Note: if `aws` was inferred incorrectly on "Acme cloud" previously, that would not be a violation.*
+- Adding new Resource attribute semantic conventions is allowed, however
+  existing resource detectors MUST NOT use them, only new detectors would be
+  allowed.
+
+If we assume capabilities of
+[Telemetry Schema conversion](https://github.com/open-telemetry/oteps/pull/161), 
+then we can relax some restrictions as follows:
```
Comment (tigrannajaryan) [inline 0162-semantic-convention-resource-stability.md:41]: So, the previous paragraph assumes no schemas? Schemas are already part of the spec and will be part of 1.4.0 spec, so I do not quite understand why we need to continue evolving the behavior in the schemaless world.

**Comment by Oberon00** [comment]
There is some related discussion on "identifying vs describing resource attributes" here: https://github.com/open-telemetry/opentelemetry-specification/issues/1298#issuecomment-775791904

**Comment by tedsuo** [comment]
@jsuereth is this step still relevant/needed?

**Comment by jsuereth** [comment]
@tedsuo The problem is relevant the solution is not. I'm closing this draft for now, as I'm crafting up a more targeted proposal for resource conventions for now.

**Pull request closed without merging**